packages/block-editor/src/components/provider/use-block-sync.js

```diff
@@ -135,6 +125,11 @@ export default function useBlockSync( {
 			setHasControlledInnerBlocks( clientId, false );
 			__unstableMarkNextChangeAsNotPersistent();
 			replaceInnerBlocks( clientId, [] );
+			// When the block becomes uncontrolled, it means its inner state has
+			// been reset we need to take the blocks again from the external
+			// value property.
+			pendingChanges.current.outgoing = [];
+			setControlledBlocks();
```

Comment:
@youknowriad I'm trying to understand if this change is fine or not. No tests are failing.

The `areInnerBlocksControlled` state is only set by the `setHasControlledInnerBlocks` action. And that action is only used twice in the code base, both in this file. It's only ever called once with `false`, which is here. So we can simply move the code here instead of using `useSelect` and a `useEffect`?

I'm not sure how to test, and I'm not sure if there's e2e test missing for this.

Reply:
Anyway, inner blocks are not controlled that often, so in terms of performance this is probably negligible. But still it's a bit simpler if possible.

Reply:
This function `unsetControlledBlocks` is only called when the block list is unmounted. I know for sure that there are some block that can change from "uncontrolled" to "controlled" while being mounted. So I think this change is probably breaking something.

That said reading the code, I wonder if we could just do this instead `const isControlled = ! clientId && value !== undefined;` to avoid the useSelect

Reply:
I forgot how to trigger it but I think the "navigation" block can move from uncontrolled to controlled while being mounted. I'm also certain that there is another use-case but can't find it right now.

Reply:
This subscription is not so important, it's only for controlled blocks anyway